spark3/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java

```diff
@@ -375,6 +376,14 @@ private void commitWithSerializableIsolation(OverwriteFiles overwriteFiles,
     private void commitWithSnapshotIsolation(OverwriteFiles overwriteFiles,
                                              int numOverwrittenFiles,
                                              int numAddedFiles) {
+      Long scanSnapshotId = scan.snapshotId();
+      if (scanSnapshotId != null) {
+        overwriteFiles.validateFromSnapshot(scanSnapshotId);
+      }
+
+      Expression conflictDetectionFilter = conflictDetectionFilter();
+      overwriteFiles.validateNoConflictingDeleteFiles(conflictDetectionFilter);
```

Comment:
Why are we check for conflicting delete files in snapshot isolation?  (I thought we must not check)

Reply:
Consider we have a data file A and there is a copy-on-write operation that overwrites it with a data file B. If a concurrent operation adds a delete file that references records from the data file A, committing the original copy-on-write operation (i.e. overwrite) would undelete the rows that were deleted concurrently.

It seems we always have to validate the delete files whenever we overwrite specific files during DELETE/MERGE.

Reply:
Got it, yea I double-checked the [Postgres doc](https://www.postgresql.org/docs/9.5/transaction-iso.html) for REPEATABLE_READ (snapshot) and it makes sense, deletes and updates to same rows must not be overridden in this mode.

Reply:
I agree. Deletes should always be validated.